### PR TITLE
nanobind: compile nanobind lib with `-DNB_FREE_THREADED` with nogil python

### PR DESCRIPTION
**Pull request by lgarrison**
Also export the flag to downstream extensions.

I tested this locally, and it seemed to have the correct behavior. Specifically, both the nanobind library and the Python extension were compiled with `-DNB_FREE_THREADED` under free-threaded Python, and neither were with "normal" Python.

As discussed in https://github.com/mesonbuild/wrapdb/issues/1758#issuecomment-2433447758.

### CODE REVIEW

**Comment by lgarrison** [comment]
CC @WillAyd

File: subprojects/packagefiles/nanobind/meson.build [hunk 1]
```diff
@@ -52,6 +52,13 @@ if get_option('buildtype') == 'release'
   add_project_arguments('-DNB_COMPACT_ASSERTIONS', language: 'cpp')
 endif
 
+freethreaded = py.get_variable('Py_GIL_DISABLED', 0) != 0
+if freethreaded
+  freethreaded_def = '-DNB_FREE_THREADED'
+  add_project_arguments(freethreaded_def, language : 'cpp')
+  dep_compile_args += [freethreaded_def]
```
Comment (WillAyd) [inline subprojects/packagefiles/nanobind/meson.build:59]: Does this need to be passed through the dependency? I was under the impression it was only needed internally by nanobind

**Review by WillAyd — COMMENTED** [review]
Thanks!

File: subprojects/packagefiles/nanobind/meson.build (hunk 1 above)
Comment (lgarrison) [reply subprojects/packagefiles/nanobind/meson.build:59]: This was a guess based on looking at the nanobind source. The main header, `nanobind/nanobind.h`, transitively includes headers like `nb_types.h` that make use of `NB_FREE_THREADED`, e.g. [here](https://github.com/wjakob/nanobind/blob/fd22b8c457cd291db2774a5526b5be39376212cd/include/nanobind/nb_types.h#L829)

Comment (WillAyd) [reply subprojects/packagefiles/nanobind/meson.build:59]: Ah interesting...I think that makes sense then

**Review by WillAyd — APPROVED** [review]
thanks @lgarrison !

**Pull request merged**